File: iOS_trip/DDSearchViewController.m
//
//  DDSearchViewController.m
//  TripDemo
//
//  Created by xiaoming han on 15/4/3.
//  Copyright (c) 2015年 AutoNavi. All rights reserved.
//

#import "DDSearchViewController.h"
#import "DDLocation.h"
#import "DDSearchManager.h"

@interface DDSearchViewController ()<UISearchBarDelegate, UISearchResultsUpdating, UITableViewDataSource, UITableViewDelegate>

@property (nonatomic, strong) UISearchController *searchController;
@property (nonatomic, strong) UITableView *tableView;
@property (nonatomic, strong) NSMutableArray *locations;

@end

@implementation DDSearchViewController

- (instancetype)init
{
    self = [super init];
    if (self)
    {
        _text = @"";
        _city = @"";
        _locations = [NSMutableArray array];
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    self.view.backgroundColor = [UIColor whiteColor];
    
    [self initTableView];
    [self initSearchBar];
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    if (self.text.length > 0)
    {
        self.searchController.searchBar.placeholder = self.text;
        [self searchTipsWithKey:self.text];
    }
    else
    {
        [self performSelector:@selector(showKeyboard) withObject:nil afterDelay:0.1];
    }
}

- (void)viewDidDisappear:(BOOL)animated
{
    [super viewDidDisappear:animated];
    self.searchController.active = NO;
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)dealloc
{
    NSLog(@"%s", __func__);
}

- (void)showKeyboard
{
    [self.searchController.searchBar becomeFirstResponder];
}

#pragma mark - Initialization

- (void)initSearchBar
{
    self.searchController = [[UISearchController alloc] initWithSearchResultsController:nil];
    
    self.searchController.searchResultsUpdater = self;
    self.searchController.dimsBackgroundDuringPresentation = NO;
    self.searchController.hidesNavigationBarDuringPresentation = NO;
    
    self.searchController.searchBar.delegate = self;
    self.searchController.searchBar.placeholder = @"请输入关键字";
    [self.searchController.searchBar sizeToFit];
    
    self.navigationItem.titleView = self.searchController.searchBar;
}

- (void)initTableView
{
    self.tableView = [[UITableView alloc] initWithFrame:CGRectMake(0, 0, self.view.frame.size.width, self.view.frame.size.height) style:UITableViewStylePlain];
    
    self.tableView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    
    self.tableView.delegate = self;
    self.tableView.dataSource = self;
    
    [self.view addSubview:self.tableView];
}

#pragma mark - Helpers

- (void)searchTipsWithKey:(NSString *)key
{
    if (key.length == 0)
    {
        return;
    }

    AMapPOIKeywordsSearchRequest *request = [[AMapPOIKeywordsSearchRequest alloc] init];
    request.requireExtension = YES;
    request.keywords = key;
    
    if (self.city.length > 0)
    {
        request.city = self.city;
    }
    
    __weak __typeof(&*self) weakSelf = self;
    [[DDSearchManager sharedInstance] searchForRequest:request completionBlock:^(id request, id response, NSError *error) {
        if (error)
        {
            NSLog(@"error :%@", error);
        }
        else
        {
            [weakSelf.locations removeAllObjects];
            
            AMapPOISearchResponse *aResponse = (AMapPOISearchResponse *)response;
            [aResponse.pois enumerateObjectsUsingBlock:^(AMapPOI *obj, NSUInteger idx, BOOL *stop)
             {
                 DDLocation *location = [[DDLocation alloc] init];
                 location.name = obj.name;
                 location.coordinate = CLLocationCoordinate2DMake(obj.location.latitude, obj.location.longitude);
                 location.address = obj.address;
                 location.cityCode = obj.citycode;
                 [weakSelf.locations addObject:location];
             }];
            
            [weakSelf.tableView reloadData];
        }
    }];
}

#pragma mark - UISearchResultsUpdating

- (void)didPresentSearchController:(UISearchController *)searchController
{
    [searchController.searchBar becomeFirstResponder];
}

- (void)updateSearchResultsForSearchController:(UISearchController *)searchController
{
    self.tableView.hidden = !searchController.isActive;
    [self searchTipsWithKey:searchController.searchBar.text];
    
    if (searchController.isActive && searchController.searchBar.text.length > 0)
    {
        searchController.searchBar.placeholder = searchController.searchBar.text;
    }
}

#pragma mark - UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.locations.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *cellIdentifier = @"cellIdentifier";
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellIdentifier];
    
    if (cell == nil)
    {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle
                                      reuseIdentifier:cellIdentifier];
    }
    
    DDLocation *location = self.locations[indexPath.row];
    
    cell.textLabel.text = location.name;
    cell.detailTextLabel.text = location.address;
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    self.searchController.active = NO;
    [tableView deselectRowAtIndexPath:indexPath animated:NO];
    
    DDLocation *location = self.locations[indexPath.row];
    if (_delegate && [_delegate respondsToSelector:@selector(searchViewController:didSelectLocation:)])
    {
        [_delegate searchViewController:self didSelectLocation:location];
    }
}

@end

File: iOS_trip-swift/DDSearchViewController.swift
//
//  DDSearchViewController.swift
//  iOS_trip
//
//  Created by liubo on 2017/1/23.
//  Copyright © 2017年 yours. All rights reserved.
//

import Foundation

protocol DDSearchViewControllerDelegate {
    func searchViewController(_ searchViewController: DDSearchViewController, didSelectLocation location: DDLocation)
}

class DDSearchViewController: UIViewController, UISearchBarDelegate, UISearchResultsUpdating, UITableViewDelegate, UITableViewDataSource {
    var delegate: DDSearchViewControllerDelegate?
    
    var text = ""
    var city = ""
    var locations = Array<DDLocation>()
    
    var searchController = UISearchController(searchResultsController: nil)
    var tableView: UITableView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = UIColor.white
        
        initTableView()
        initSearchBar()
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        if text.characters.count > 0 {
            searchController.searchBar.placeholder = text
            searchTipsWithKey(text)
        }
        else {
            perform(#selector(self.showKeyboard), with: nil, afterDelay: 0.1)
        }
    }
    
    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        searchController.isActive = false
    }
    
    func showKeyboard() {
        searchController.searchBar.becomeFirstResponder()
    }
    
    //MARK: - Initialization
    
    func initTableView() {
        tableView = UITableView(frame: CGRect(x: 0, y: 0, width: view.frame.size.width, height: view.frame.size.height), style: .plain)
        tableView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        tableView.delegate = self
        tableView.dataSource = self
        
        view.addSubview(tableView)
    }
    
    func initSearchBar() {
        searchController.searchResultsUpdater = self
        searchController.dimsBackgroundDuringPresentation = false
        searchController.hidesNavigationBarDuringPresentation = false
        
        searchController.searchBar.delegate = self
        searchController.searchBar.placeholder = "请输入关键字"
        searchController.searchBar.sizeToFit()
        
        navigationItem.titleView = searchController.searchBar
    }
    
    //MARK: - Helpers
    
    func searchTipsWithKey(_ key: String) {
        if key.characters.count == 0 {
            return
        }
        
        let request = AMapPOIKeywordsSearchRequest()
        request.requireExtension = true
        request.keywords = key
        
        if self.city.characters.count > 0 {
            request.city = self.city
        }
        
        DDSearchManager.Shared.searchForRequest(request: request, completionBlock: {[weak self] (request: AMapSearchObject, response: AMapSearchObject?, error: NSError?) in
            if let error = error {
                let error = error as NSError
                NSLog("error :%@", error)
                return
            }
            else {
                self?.locations.removeAll()
                
                guard let aResponse = response as? AMapPOISearchResponse else {
                    return
                }
                
                for obj in aResponse.pois {
                    let location = DDLocation()
                    location.name = obj.name
                    location.coordinate = CLLocationCoordinate2DMake(CLLocationDegrees(obj.location.latitude), CLLocationDegrees(obj.location.longitude))
                    location.address = obj.address
                    location.cityCode = obj.citycode
                    self?.locations.append(location)
                }
                
                self?.tableView.reloadData()
            }
        })
    }
    
    //MARK: - UISearchResultsUpdating
    
    func didPresentSearchController(_ searchController: UISearchController) {
        searchController.searchBar.becomeFirstResponder()
    }
    
    func updateSearchResults(for searchController: UISearchController) {
        tableView.isHidden = !searchController.isActive
        
        guard let str = searchController.searchBar.text else {
            return
        }
        
        searchTipsWithKey(str)
        
        if searchController.isActive && str.characters.count > 0 {
            searchController.searchBar.placeholder = str
        }
    }
    
    //MARK: - UITableViewDataSource
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return locations.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cellIdentifier = "cellIdentifier"
        
        var cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier)
        if cell == nil {
            cell = UITableViewCell(style: .subtitle, reuseIdentifier: cellIdentifier)
        }
        
        let location = locations[indexPath.row]
        
        cell?.textLabel?.text = location.name
        cell?.detailTextLabel?.text = location.address
        
        return cell!
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        searchController.isActive = false
        tableView.deselectRow(at: indexPath, animated: false)
        
        let location = locations[indexPath.row]
        if delegate != nil {
            delegate?.searchViewController(self, didSelectLocation: location)
        }
    }
}
